include/dgl/array.h

```diff
@@ -57,30 +105,169 @@ IdArray Sub(dgl_id_t lhs, IdArray rhs);
+ * The column indices are 0-based and are not necessarily sorted.
+ *
+ * Note that we do allow duplicate non-zero entries -- multiple non-zero entries
+ * that have the same row, col indices. It corresponds to multigraph in
+ * graph terminology.
+ */
 struct CSRMatrix {
-  IdArray indptr, indices, data;
+  /*! \brief the dense shape of the matrix */
+  int64_t num_rows, num_cols;
+  /*! \brief CSR index arrays */
+  runtime::NDArray indptr, indices;
+  /*! \brief data array, could be empty. */
+  runtime::NDArray data;
 };
 
-/*! \brief Plain COO structure */
+/*!
+ * \brief Plain COO structure
+ * 
+ * Note that we do allow duplicate non-zero entries -- multiple non-zero entries
+ * that have the same row, col indices. It corresponds to multigraph in
+ * graph terminology.
+ *
+ * We call a COO matrix is *coalesced* if its row index is sorted.
+ */
 struct COOMatrix {
-  IdArray row, col, data;
+  /*! \brief the dense shape of the matrix */
+  int64_t num_rows, num_cols;
+  /*! \brief COO index arrays */
+  runtime::NDArray row, col;
+  /*! \brief data array, could be empty. */
+  runtime::NDArray data;
 };
 
-/*! \brief Slice rows of the given matrix and return. */
-CSRMatrix SliceRows(const CSRMatrix& csr, int64_t start, int64_t end);
+///////////////////////// CSR routines //////////////////////////
+
+/*! \brief Return true if the value (row, col) is non-zero */
+bool CSRIsNonZero(CSRMatrix , int64_t row, int64_t col);
+runtime::NDArray CSRIsNonZero(CSRMatrix, runtime::NDArray row, runtime::NDArray col);
 
-/*! \brief Convert COO matrix to CSR matrix. */
-CSRMatrix ToCSR(const COOMatrix);
+/*! \brief Return the nnz of the given row */
+int64_t CSRGetRowNNZ(CSRMatrix , int64_t row);
+runtime::NDArray CSRGetRowNNZ(CSRMatrix , runtime::NDArray row);
+
+/*! \brief Return the column index array of the given row */
+runtime::NDArray CSRGetRowColumnIndices(CSRMatrix , int64_t row);
+
+/*! \brief Return the data array of the given row */
+runtime::NDArray CSRGetRowData(CSRMatrix , int64_t row);
+
+/* \brief Get data. The return type is an ndarray due to possible duplicate entries. */
+runtime::NDArray CSRGetData(CSRMatrix , int64_t row, int64_t col);
+runtime::NDArray CSRGetData(CSRMatrix, runtime::NDArray rows, runtime::NDArray cols);
+
+/* \brief Get the data and the row,col indices for each returned entries. */
+std::vector<runtime::NDArray> CSRGetDataAndIndices(
+    CSRMatrix , runtime::NDArray rows, runtime::NDArray cols);
+
+/*! \brief Return a transposed CSR matrix */
+CSRMatrix CSRTranspose(CSRMatrix csr);
+
+/*!
+ * \brief Convert COO matrix to CSR matrix.
+ * \param csr Input csr matrix
+ * \param data_as_order If true, the data array in the input csr matrix contains the order
+ *                      by which the resulting COO tuples are stored. In this case, the
+ *                      data array of the resulting COO matrix will be empty because it
+ *                      is essentially a consecutive range.
+ * \return a coo matrix
```

Comment:
The `data_as_order` seems to be a very special case. When do we have such scenario?